cli/src/cmd/cast/bind.rs

```diff
@@ -0,0 +1,148 @@
+use forge::Address;
+use foundry_common::fs::json_files;
+use futures::future::BoxFuture;
+
+use crate::{cmd::Cmd, opts::ClapChain};
+
+static DEFAULT_CRATE_NAME: &str = "foundry-contracts";
+static DEFAULT_CRATE_VERSION: &str = "0.0.1";
+
+#[derive(Debug, Clone, Parser)]
+pub struct BindArgs {
+    #[clap(
+        help = "The contract address, or the path to an ABI Directory.",
+        long_help = r#"The contract address, or the path to an ABI Directory.
+
+If an address is specified, then the ABI is fetched from Etherscan."#,
+        value_name = "PATH_OR_ADDRESS"
+    )]
+    path_or_address: String,
+
+    #[clap(long, short, env = "ETHERSCAN_API_KEY", help = "etherscan API key", value_name = "KEY")]
+    etherscan_api_key: Option<String>,
+    #[clap(
+        help = "Path to where bindings will be stored",
+        long = "output-dir",
+        short,
+        value_hint = ValueHint::DirPath,
+        value_name = "PATH"
+    )]
+    pub output_dir: Option<PathBuf>,
+
+    #[clap(
+        long = "crate-name",
+        help = "The name of the Rust crate to generate. This should be a valid crates.io crate name. However, it is not currently validated by this command.",
+        default_value = DEFAULT_CRATE_NAME,
+    )]
+    crate_name: String,
+
+    #[clap(
+        long = "crate-version",
+        help = "The version of the Rust crate to generate. This should be a standard semver version string. However, it is not currently validated by this command.",
+        default_value = DEFAULT_CRATE_VERSION,
+        value_name = "NAME"
+    )]
+    crate_version: String,
+
+    #[clap(long = "seperate-files", help = "Generate bindings as seperate files.")]
+    seperate_files: bool,
+
+    #[clap(flatten)]
+    chain: ClapChain,
+}
+
+impl Cmd for BindArgs {
+    type Output = BoxFuture<'static, eyre::Result<()>>;
+
+    fn run(self) -> eyre::Result<Self::Output> {
+        let cmd = Box::pin(async move {
+            let bind_args = self.clone();
+            if Path::new(&self.path_or_address).exists() {
+                self.generate_bindings(AbiPath::Local {
+                    path: bind_args.path_or_address,
+                    name: None,
+                })
+                .await
+            } else {
+                self.generate_bindings(AbiPath::Etherscan {
+                    address: bind_args.path_or_address.parse::<Address>().unwrap(),
+                    chain: bind_args.chain.inner,
+                    api_key: bind_args.etherscan_api_key.unwrap(),
+                })
+                .await
+            }
+        });
+
+        Ok(cmd)
+    }
+}
+
+impl BindArgs {
+    pub async fn generate_bindings(&self, address_or_path: AbiPath) -> eyre::Result<()> {
+        match address_or_path {
+            AbiPath::Etherscan { address, chain, api_key } => {
+                let client = Client::new(chain, api_key)?;
+                let source = match client.contract_source_code(address).await {
```

Comment:
pls add this logic to also make this work with proxies

https://github.com/foundry-rs/foundry/pull/484/files#diff-2a68e14bf95d2e3c7c7c72b92741c10dc2ea9027388a899a65a9a73b4152b80fR192-R197